Adds the VAEEncoder to StableDiffusion

LukeWood opened the pull request:
TODOS:

- [x] port and upload weights
- [x] figure out if padding needs to be done to non-512x512 images

fchollet reviewed (commented):
Thanks for the PR!

LukeWood commented on keras_cv/models/generative/stable_diffusion/vae_encoder.py:
```diff
@@ -0,0 +1,56 @@
+    def __init__(self):
+        super().__init__(
+            [
+                PaddedConv2D(128, 3, padding=1),
```
Done

LukeWood commented:
@fchollet added weight loading.  Any review comments?

LukeWood commented:
/gcbrun

LukeWood commented:
Weird; the VAEEncoder weights are not loading:

```
>>> model = keras_cv.models.StableDiffusion()
Downloading data from https://huggingface.co/fchollet/stable-diffusion/blob/main/vae_encoder.h5
36499/36499 [==============================] - 0s 2us/step
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lukewood/workspace/keras-cv/keras_cv/models/generative/stable_diffusion/stable_diffusion.py", line 124, in __init__
    self.image_encoder.load_weights(image_encoder_weights_fpath)
  File "/usr/local/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 67, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/usr/local/lib/python3.9/site-packages/h5py/_hl/files.py", line 424, in __init__
    fid = make_fid(name, mode, userblock_size,
  File "/usr/local/lib/python3.9/site-packages/h5py/_hl/files.py", line 190, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 96, in h5py.h5f.open
OSError: Unable to open file (file signature not found)
```

I've tried deleting and re-downloading but still get the error.

bhack commented:
> ttps://huggingface.co/fchollet/stable-diffusion/blob/main/

Have you verified `SHA256: 7ae08f918d4e3a686807474706a6821ca1ff3c71b84d98ba33081e6cb7aab9e3`?

bhack commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -107,9 +113,15 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
         )
+        image_encoder_weights_fpath = keras.utils.get_file(
+            origin="https://huggingface.co/fchollet/stable-diffusion/blob/main/vae_encoder.h5",
+            file_hash="005390320d6095734833d0b442d11f413aae42c4d42872f3b352f7f53b5da122",
```
it is now SHA256: 7ae08f918d4e3a686807474706a6821ca1ff3c71b84d98ba33081e6cb7aab9e3

bhack replied:
> OSError: Unable to open file (file signature not found)
The error message has a very low usability.

LukeWood replied:
Trying now with a manually specified hash algorithm.

bhack replied:
https://huggingface.co/fchollet/stable-diffusion/blob/main/vae_encoder.h5
![immagine](https://user-images.githubusercontent.com/1710528/193895372-67662b2b-8cfe-42e0-af1f-d5f627d86ee0.png)

bhack replied:
It doesn't match locally with wget:

```bash
sha256sum vae_encoder.h5 
c60fb220a40d090e0f86a6ab4c312d113e115c87c40ff75d11ffcf380aab7ebb`

md5sum vae_encoder.h5 
d3c5e2b8e4eee794417f9552fe0cb388  vae_encoder.h5
```

LukeWood replied:
Updated to work.  try clearing your python cache

bhack replied:
> Updated to work. try clearing your "~/.keras" cache

It is not the cache, It was commited `c60` 14 minutes ago..

LukeWood replied:
I know.  I just mean't try re-running now to confirm it works.

LukeWood commented:
ok @fchollet this now works!

LukeWood commented:
/gcbrun

LukeWood commented:
/gcbrun

fchollet commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -89,10 +90,15 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
         self.decoder = Decoder(img_height, img_width)
+
+        # TODO(lukewood): should we support variable size image inputs here?
+        self.image_encoder = VAEEncoder()
```
This model will only be used for some use cases but creating it here will induce time and RAM overhead for all users. We should switch to a lazy model creation setup imo, something like `create_image_encoder()` called at the start of the method that requires it

LukeWood replied:
we'll likely just use the image encoder directly in the tutorial.  Perhaps a lazy getter would work?

Re: `ImageEncoder` SGTM.

fchollet commented on keras_cv/models/generative/stable_diffusion/vae_encoder.py:
```diff
@@ -0,0 +1,60 @@
+    def __init__(self):
+        super().__init__(
+            [
+                keras.layers.Input((512, 512, 3)),
```
Please make the image dimensions keyword arguments, passed to be the user-provided values when creating the model

LukeWood commented:
@fchollet what do you think about exposing the sub-models as APIs?  This would make writing TextualInversion a lot easier.  The one issue is the text models are very much NLP models.

LukeWood commented:
Addressed all comments, PTAL.  My plan for now is to host the auxiliary modules needed in a 3p repo for now to get the guide written, then we can look at the best way to expose these APIs in a more permanent manner.

LukeWood commented:
@fchollet let me know if you see any issues here.  I'm planning a pass on the Textual Inversion tutorial this weekend

ianstenbit commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -77,7 +82,14 @@ class StableDiffusion:
+        img_height=512,
+        img_width=512,
+        input_img_height=512,
+        input_img_width=512,
```
I don't love stacking these arguments here, but I also think asking for an image encoder in the constructor is a bit weird.

I suppose we could add a method `addImageEncoder` or `withImageEncoder`, but that feels a bit awkward also.

ianstenbit replied:
How would a user initialize StableDiffusion with custom pre-trained weights for the VAE?

LukeWood replied:
@fchollet any thoughts on this?  I also dont love stacking the arguments like this.  Perhaps we just assume the input & output images are of static shape?

Regarding custom weights: you don't.  StableDiffusion is only mean't to replicate exact numerics of the checkpoint; nothing more, nothing less.

fchollet replied:
Why not use the same size across the board?

LukeWood replied:
I guess theres no use case YET to use a dynamic input shape; but perhaps there may be one at some point?  

In any case, I have modified this.

ianstenbit commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -89,10 +101,16 @@ def __init__(self, img_height=512, img_width=512, jit_compile=False):
             self.text_encoder.compile(jit_compile=True)
             self.diffusion_model.compile(jit_compile=True)
             self.decoder.compile(jit_compile=True)
+            self.image_encoder.compile(jit_compile=True)
```
If I understand correctly, this causes us to instantiate an image encoder model whenever jit_compile is being used (regardless of whether it's actually going to be used). I think this is probably not the behavior we want.

LukeWood replied:
Oh yes, that is correct.  Good catch!

ianstenbit commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -281,6 +300,14 @@ def _get_unconditional_context(self):
         return unconditional_context
 
+    @property
+    def image_encoder(self):
```
I think if we're adding an image_encoder property, we may also want an API surface to use it (e.g. `encode_image`)?

LukeWood replied:
I was thinking that you just run `stable_diffusion.image_encoder.predict()`.  

Perhaps this is not a great API surface though?

fchollet replied:
Seems fine!

ianstenbit commented on keras_cv/models/generative/stable_diffusion/image_encoder.py:
```diff
@@ -0,0 +1,67 @@
+    ResnetBlock,
+)
+
+
```
I think a docstring for this encoder would probably be useful (at least to explain that it's a VAE and a reference)

LukeWood commented:
/gcbrun

LukeWood commented:
ok @fchollet @ianstenbit I think we're ready to merge!

ianstenbit commented on keras_cv/models/generative/stable_diffusion/__internal__/layers/attention_block.py:
```diff
@@ -0,0 +1,56 @@
+# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
```
KerasCV authors

ianstenbit commented on keras_cv/models/generative/stable_diffusion/__internal__/layers/resnet_block.py:
```diff
@@ -0,0 +1,43 @@
+)
+
+
+class ResnetBlock(keras.layers.Layer):
```
nit: should this be ResNetBlock?

LukeWood replied:
Lets leave this as a potential cleanup - this PR just moves the block but didn't add it.

ianstenbit commented on keras_cv/models/generative/stable_diffusion/stable_diffusion.py:
```diff
@@ -281,6 +293,14 @@ def _get_unconditional_context(self):
         return unconditional_context
 
+    @property
+    def image_encoder(self):
```
Maybe worth a docstring pointing to the ImageEncoder class or showing an example of its use?

LukeWood replied:
Good point - added this!

LukeWood commented:
/gcbrun

LukeWood commented:
Thanks @ianstenbit and @fchollet for the review, excited to use this in the TextualInversion guide!

fchollet reviewed (commented):
LGTM
